### PR TITLE
Support returning error values instead of raising in `get` and `cas`

### PR DESCRIPTION
**Pull request by catlee**
I'd like to experiment with using return values from `get` / `cas` instead of raising exceptions.

I focused on these two methods for now since they seem to be in the critical path of most operations. We could potentially expand this approach out to other methods, and also representing the different regular return states (success, data exists), of `cas` properly.

### CODE REVIEW

File: lib/memcached/memcached.rb [hunk 1]
```diff
@@ -453,6 +460,9 @@ def cas(keys, ttl=@default_ttl, decode=true)
     else
       # Single CAS
       value, flags, cas = single_get(keys, decode)
+      if !@raise_on_errors && value.equal?(NOT_FOUND)
```
Comment (casperisfine) [inline lib/memcached/memcached.rb]:
```suggestion
      if !@raise_on_errors && NOT_FOUND == value
```

micro-optimization: you'd think `.equal?` would be the fastest thing ever, but actually it is slower than `==` in many cases because it doesn't have a dedicated `opt_` opcode like `==` does.

File: test/unit/memcached_test.rb [hunk 2]
```diff
@@ -6,6 +6,13 @@ def to_str
   end
 end
 
+# assert_equal calls obj.respond_to?, which isn't supported by NOT_FOUND < BasicObject
+class Memcached
+  def NOT_FOUND.respond_to?(*)
```
Comment (casperisfine) [inline test/unit/memcached_test.rb]: This makes me a bit nervous, because we could end up calling it in the library code and not realize it's broken.

File: lib/memcached/memcached.rb [hunk 3]
```diff
@@ -585,6 +595,15 @@ def stats(subcommand = nil)
...
   end
 
+  # Causes get / cas operations to return NOT_FOUND instead of raising NotFound
+  def without_exceptions(&block)
+    prev = @raise_on_errors
+    @raise_on_errors = false
+    yield
+  ensure
+    @raise_on_errors = prev
+  end
```
Comment (casperisfine) [inline lib/memcached/memcached.rb:609]: Is this really needed? I'm assuming this feature will be used by `memcached_store`, so making the whole store compatible doesn't seem too hard.

File: test/unit/memcached_test.rb (hunk 2 above)
Comment (casperisfine) [reply test/unit/memcached_test.rb]: I took the liberty to push a commit that moves this gem to `minitest` removing the need for that hack.

File: lib/memcached/memcached.rb [hunk 4]
```diff
@@ -738,9 +760,10 @@ def multi_cas(hash, ttl, flags_and_cas, encode, tries)
...
   def single_cas(key, value, ttl, flags, cas, encode)
     value, flags = @codec.encode(key, value, flags) if encode
-    check_return_code(
-      Lib.memcached_cas(memcached_struct, key, value, ttl, flags, cas),
-      key
-    )
+    ret = Lib.memcached_cas(memcached_struct, key, value, ttl, flags, cas)
+    if ret == 16 && !@raise_on_errors
+      return NOT_FOUND
+    end
```
Comment (casperisfine) [inline lib/memcached/memcached.rb]: I think it would make more sense to move this in `check_return_code`

File: lib/memcached/memcached.rb (hunk 3 above)
Comment (catlee) [reply lib/memcached/memcached.rb:609]: I wanted a way to enable this behaviour selectively before committing to using it everywhere.

Comment (casperisfine) [reply lib/memcached/memcached.rb:609]: Yeah, I get that, but we might as well do the full sweep.

File: lib/memcached/memcached.rb [hunk 5]
```diff
@@ -585,18 +589,37 @@ def stats(subcommand = nil)
...
+      if @raise_on_errors
+        raise NotFound, "NOTFOUND".freeze, @show_backtraces, cause: nil # Lib::MEMCACHED_NOTFOUND
+      else
+        NOT_FOUND
+      end
+    when 14
+      if @raise_on_errors
+        raise NotStored, "NOTSTORED".freeze, @show_backtraces, cause: nil # Lib::MEMCACHED_NOTSTORED
+      else
+        NOT_FOUND
```
Comment (catlee) [inline lib/memcached/memcached.rb]: If we're going to try and handle all normal cases, we should return `NOT_STORED` here so that callers can distinguish between these cases, and then also handle 12 -> `DATA_EXISTS`, which can be returned from `cas` if the data has been changed since reading it.

File: lib/memcached/memcached.rb (hunk 3 above)
Comment (camilo) [reply lib/memcached/memcached.rb:609]: @casperisfine are you suggesting to completely remove the exception logic instead of adding the new method? I think so, but just to be sure

File: test/unit/memcached_test.rb [hunk 6]
```diff
@@ -22,6 +22,7 @@ def setup
       :distribution => :modula,
       :show_backtraces => true}
     @cache = Memcached.new(@servers, @options)
+    @cache.flush
```
Comment (camilo) [inline test/unit/memcached_test.rb:25]: why is this needed? precaution in case there is garbage on the backend?

File: test/unit/memcached_test.rb [hunk 7]
```diff
@@ -1279,6 +1282,7 @@ def test_unresponsive_server_retries_equals_server_failure_limit
   end
 
   def test_unresponsive_server_retries_less_than_server_failure_limit
+    skip "This test is broken"
```
Comment (camilo) [inline test/unit/memcached_test.rb]: how is it broken?

File: lib/memcached/memcached.rb [hunk 8]
```diff
@@ -585,18 +599,46 @@ def stats(subcommand = nil)
...
+  ensure
+    @raise_on_errors = prev
+  end
+
   ### Operations helpers
 
   private
 
   # Checks the return code from Rlibmemcached against the exception list. Raises the corresponding exception if the return code is not Memcached::Success or Memcached::ActionQueued. Accepts an integer return code and an optional key, for exception messages.
+  # When @raise_on_errors is false the following are returned instead of raiing exceptions:
```
Comment (camilo) [inline lib/memcached/memcached.rb]: nit: s/raiing/raising/

File: lib/memcached/memcached.rb [hunk 9]
```diff
@@ -585,18 +599,46 @@ def stats(subcommand = nil)
...
+  # `NOT_STORED` is returned when ret == 12 (Lib::MEMCACHED_DATA_EXISTS)
+  # `NOT_FOUND` is returned when ret == 16 (Lib::MEMCACHED_NOTFOUND)
+  # `DATA_EXISTS` is returned when ret == 14 (Lib::MEMCACHED_NOTSTORED)
   def check_return_code(ret, key = nil) #:doc:
```
Comment (camilo) [inline lib/memcached/memcached.rb:620]: I see where @casperisfine is going. This function feels a bit awkward now, if we dropped the exceptions completely (not sure how much work that is) then it'd be called something like  `return_code` and just give back a value instead of the nil check. ( I guess we could say now is a bit golang inspired `if err != nil` )

File: lib/memcached/memcached.rb (hunk 3 above)
Comment (casperisfine) [reply lib/memcached/memcached.rb:609]: Not remove it entirely, but just to have a `MemcachedStore` that always check the return value and never rescue NOT_FOUND.

I don't think it's much more work to handle all the codepaths so might as well.

File: test/unit/memcached_test.rb (hunk 6 above)
Comment (casperisfine) [reply test/unit/memcached_test.rb:25]: Because I was running into leaky tests. They predate this PR, I just fixed it there.

File: test/unit/memcached_test.rb (hunk 7 above)
Comment (casperisfine) [reply test/unit/memcached_test.rb]: I have no idea, it's trying to assert something that makes little sense. I fixed the one above it, but this one I couldn't figure out.

Comment (casperisfine) [reply test/unit/memcached_test.rb]: Note that it's broken before this PR.

Comment (camilo) [reply test/unit/memcached_test.rb]: >Note that it's broken before this PR.

okay did not realize as it seems that the skip is new

Comment (casperisfine) [reply test/unit/memcached_test.rb]: The skip is new. The test only passes if another test is ran before:

```
(1-0-stable-shopify)$ bundle exec ruby -Itest test/unit/memcached_test.rb -n test_unresponsive_server_retries_less_than_server_failure_limit
Loaded suite test/unit/memcached_test
Started
F
============================================================================================================================================================================================================================================================================================================================
Failure: test_unresponsive_server_retries_less_than_server_failure_limit(MemcachedTest)
test/unit/memcached_test.rb:1300:in `rescue in test_unresponsive_server_retries_less_than_server_failure_limit'
test/unit/memcached_test.rb:1297:in `test_unresponsive_server_retries_less_than_server_failure_limit'
     1294: 
     1295:     key2 = 'test_missing_server'
     1296:     assert_raise(Memcached::ATimeoutOccurred) { cache.set(key2, @value) }
  => 1297:     begin
     1298:       cache.get(key2)
     1299:     rescue => e
     1300:       assert_equal Memcached::ServerIsMarkedDead, e.class
<Memcached::ServerIsMarkedDead> expected but was
<Memcached::NotFound>

diff:
? Memcached::ServerIsMarkedDead
?            NotFoun           
============================================================================================================================================================================================================================================================================================================================

Finished in 1.029725 seconds.
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
1 tests, 2 assertions, 1 failures, 0 errors, 0 pendings, 0 omissions, 0 notifications
0% passed

```

The addition of the `flush` call and introduction of random ordering is what broke it in this PR.

I could configure minitest to run tests in order and remove the `flush` but I think it is a worst outcome.

Comment (casperisfine) [reply test/unit/memcached_test.rb]: Ok, I dug into it and fixed it in https://github.com/Shopify/memcached/pull/28/commits/7653e7319a07a21046de5ecdf9bdb1dd303f338c
